### PR TITLE
feat(assistantV2): message[][] data model

### PR DESCRIPTION
**Pull request by fontanierh**
Early data model, only a subset of the tables and columns.

WIP of a more complete data model [here](https://dbdiagram.io/d/64f0bb5702bd1c4a5ec22505)

The 2 main things:
## `Message` is "polymorphic"

it can be either a `UserMessage` or an `AssistantMessage`. This is done by having 2 nullable foreign keys (one to each table) and enforcing that exactly one of them must be non-null.

## the `[][]` part

the fact that messages in the conversation are append only but can be edited / retried is implemented on the `Message` table using a unique index on columns (`version`, `conversation_id`, `rank`). For each rank, the "current" message is the one with the highest `version`. This allows to display previous versions of a specific message in the right order.
Editing a message / regenerating a response means creating a new `Message` with the same `(rank, conversation_id)` but with an incremented `version`.

### CODE REVIEW

File: front/lib/models.ts [hunk 1]
```diff
@@ -1241,6 +1241,189 @@ DataSource.hasMany(DocumentTrackerChangeSuggestion, {
...
+    modelName: "message",
+    sequelize: front_sequelize,
+    indexes: [
+      {
+        unique: true,
+        fields: ["version", "conversation_id", "rank"],
+      },
+    ],
+    hooks: {
+      // TODO @fontanierh: check if we want to add a Check Constraint (from db.ts ?)
```
Comment (fontanierh) [inline front/lib/models.ts:1397]: @spolu  Do we want to enforce this at DB level or is ORM enough ?
Sequelize doesn't support adding the `CheckConstraint`, but it's easy enough to add with raw SQL in `db.ts`.

File: front/lib/models.ts [hunk 2]
```diff
@@ -1241,6 +1241,194 @@ DataSource.hasMany(DocumentTrackerChangeSuggestion, {
...
+}
+
+Message.init(
+  {
+    id: {
+      type: DataTypes.INTEGER,
+      autoIncrement: true,
+      primaryKey: true,
+    },
+    parentId: {
```
Comment (fontanierh) [inline front/lib/models.ts:1354]: @spolu  Didn't make this unique, so making the assumption that multiple messages can have the same parent. WDYT ?

File: front/lib/models.ts [hunk 3]
```diff
@@ -1241,6 +1241,183 @@ DataSource.hasMany(DocumentTrackerChangeSuggestion, {
...
+});
+Message.belongsTo(UserMessage, { foreignKey: "userMessageId" });
+AssistantMessage.hasOne(Message, {
+  foreignKey: "assistantMessageId",
+  sourceKey: "id",
+  onDelete: "RESTRICT",
+});
+Message.belongsTo(AssistantMessage, { foreignKey: "assistantMessageId" });
+Message.belongsTo(Message, { foreignKey: "parent_id", as: "parent" });
+Message.hasMany(Message, { foreignKey: "parent_id", as: "children" });
```
Comment (fontanierh) [inline front/lib/models.ts:1419]: @spolu
This foreign key isn't unique so making the assumption that multiple messages can have the same parent. WDYT ?

File: front/lib/models.ts [hunk 4]
```diff
@@ -1241,6 +1241,187 @@ DataSource.hasMany(DocumentTrackerChangeSuggestion, {
...
+  InferAttributes<Message>,
+  InferCreationAttributes<Message>
+> {
+  declare id: number;
+
+  declare conversation_id: number;
+  declare version: number;
+  declare rank: number;
+
+  declare is_deleted: boolean;
```
Comment (fontanierh) [inline front/lib/models.ts:1347]: @spolu  Is there a strong reason to prefer `status: "visible" | "deleted"` over a boolean ?

File: front/lib/models.ts (hunk 1 above)
Comment (spolu) [reply front/lib/models.ts:1403]: I'm fine not enforcing for now

File: front/lib/models.ts (hunk 4 above)
Comment (spolu) [reply front/lib/models.ts:1355]: My 5 years at Stripe has taught me on thing: always status never a boolean

File: front/lib/models.ts (hunk 3 above)
Comment (spolu) [reply front/lib/models.ts:1436]: Yes @A @B => multiple message will have the same parent indeed

**Review by spolu — COMMENTED** [review]
Let's rename as discussed IRL

Product is Assistant
Thingy is Agent
Other thingy is Conversation

File: front/lib/models.ts [hunk 5]
```diff
@@ -1241,6 +1241,200 @@ DataSource.hasMany(DocumentTrackerChangeSuggestion, {
...
+}
+
+UserMessage.init(
+  {
+    id: {
+      type: DataTypes.INTEGER,
+      autoIncrement: true,
+      primaryKey: true,
+    },
+    textContent: {
```
Comment (spolu) [inline front/lib/models.ts:1304]: would call it message
we also need to add:

userContextUsername (required)
userContextTimezone (required)
userContextFullName (optional)
userContenxtEmail (optional)
userContextProfilePictureUrl (optional)

Not sure if we want to linearize everything or use a postgres JSON for that?

**Comment by spolu** [comment]
Bullish on cutting these in their own file BTW (if it's an easy one)

**Comment by fontanierh** [comment]
> Bullish on cutting these in their own file BTW (if it's an easy one)

aligned

**Comment by fontanierh** [comment]
@spolu  big diff but it's only the split models file commit

File: front/lib/models/assistant.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+      primaryKey: true,
+    },
+    sId: {
+      type: DataTypes.STRING,
+      allowNull: false,
+      unique: true,
+    },
+    title: {
+      type: DataTypes.TEXT,
+      allowNull: false,
```
Comment (spolu) [inline front/lib/models/assistant.ts]: allowNull should be true

File: front/lib/models/assistant.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+  declare assistantConversationId: ForeignKey<AssistantConversation["id"]>;
+  declare parentId: ForeignKey<AssistantMessage["id"]> | null;
+  declare assistantUserMessageId: ForeignKey<AssistantUserMessage["id"]> | null;
+  declare assistantAgentMessageId: ForeignKey<
+    AssistantAgentMessage["id"]
+  > | null;
+}
+
+AssistantMessage.init(
```
Comment (spolu) [inline front/lib/models/assistant.ts:159]: Should be above next to Conversation

File: front/lib/models/assistant.ts [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    },
+  }
+);
+
+AssistantMessage.belongsTo(AssistantConversation, {
+  foreignKey: "assistantConversationId",
+  onDelete: "CASCADE",
+});
+AssistantConversation.hasMany(AssistantMessage, {
+  foreignKey: "assistantAgentMessageId",
```
Comment (spolu) [inline front/lib/models/assistant.ts:216]: wrong foreignKey no?

File: front/lib/models/assistant.ts [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+});
+AssistantUserMessage.hasOne(AssistantMessage, {
+  foreignKey: "assistantUserMessageId",
+  sourceKey: "id",
+  onDelete: "RESTRICT",
```
Comment (spolu) [inline front/lib/models/assistant.ts:221]: what is RESTRICT?

File: front/lib/models/assistant.ts [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+});
+AssistantMessage.belongsTo(AssistantUserMessage, {
```
Comment (spolu) [inline front/lib/models/assistant.ts:223]: why?

Comment (spolu) [reply front/lib/models/assistant.ts:223]: feels inverted?

File: front/lib/models/assistant.ts [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+  onDelete: "RESTRICT",
+});
+AssistantMessage.belongsTo(AssistantAgentMessage, {
+  foreignKey: "assistantAgentMessageId",
+});
+AssistantMessage.belongsTo(AssistantMessage, {
+  foreignKey: "parentId",
+  as: "parent",
+});
+AssistantMessage.hasMany(AssistantMessage, {
```
Comment (spolu) [inline front/lib/models/assistant.ts:238]: lots of magic needs commenting

**Review by spolu — CHANGES_REQUESTED** [review]

File: front/lib/models/assistant.ts (hunk 6 above)
Comment (fontanierh) [reply front/lib/models/assistant.ts]: updated

File: front/lib/models/assistant.ts (hunk 8 above)
Comment (fontanierh) [reply front/lib/models/assistant.ts:216]: indeed

**Review by spolu — COMMENTED** [review]
Let's try to have the minimum set of sequelize "realation" verbs

File: front/lib/models/assistant.ts (hunk 9 above)
Comment (fontanierh) [reply front/lib/models/assistant.ts]: Removed it, we'll use whatever is default here

File: front/lib/models/assistant.ts (hunk 7 above)
Comment (fontanierh) [reply front/lib/models/assistant.ts:158]: As per IRL, it has foreign keys to the other ones so might be clearer if below

File: front/lib/models/assistant.ts (hunk 10 above)
Comment (fontanierh) [reply front/lib/models/assistant.ts]: Removed all the belongsto

File: front/lib/models/assistant.ts (hunk 11 above)
Comment (fontanierh) [reply front/lib/models/assistant.ts:218]: Cleaned up quite a bit of those, should be clear enough w/o comments now I think ? lmk

**Comment by fontanierh** [comment]
@spolu  cleaned up

File: front/lib/models/assistant.ts (hunk 11 above)
Comment (spolu) [reply front/lib/models/assistant.ts]: let's do belongsTo foreignKey: parentId

**Review by spolu — DISMISSED** [review]
LGTM

**Review by spolu — APPROVED** [review]

**Pull request merged**